tools/gen-unicode.py

```diff
@@ -21,275 +21,264 @@
-        #   connector punctuation: Pc
-        # separators:              Zs
-        self._unicode_categories = {
-            'letters_category' : ["Lu", "Ll", "Lt", "Lm", "Lo", "Nl"],
-            'non_letters_category' : ["Nd", "Mn", "Mc", "Pc"],
-            'separators_category' : ["Zs"]
-        }
-
-        self._categories = {
-            'letters' : [],
-            'non_letters' : [],
-            'separators' : []
-        }
-
-    def _store_by_category(self, unicode_id, category):
+        self._length_limit = 0xff
+        self.extra_id_continue_units = set([0x200C, 0x200D])
+
+    def in_range(self, i):
+        return i >= 0x80 and i < 0x10000
+
+    def _group_ranges(self, units):
         """
-        Store the given unicode_id by its category
+        Convert an increasing list of integers into a range list
+        :return: List of ranges.
         """
-        for target_category in self._categories:
-            if category in self._unicode_categories[target_category + '_category']:
-                self._categories[target_category].append(unicode_id)
+        for _, group in itertools.groupby(enumerate(units), lambda q: (q[1] - q[0])):
+            group = list(group)
+            yield group[0][1], group[-1][1]
 
-    def read_categories(self, unicode_data_file):
-        """
-        Read the corresponding unicode values and store them in category lists.
 
-        :return: List of letters, non_letter and separators.
+    def create_tables(self, units):
+        """
+        Split list of ranges into intervals and single char lists.
+        :return: List of interval starting points, interval lengths and single chars
         """
 
-        range_start_id = 0
-
-        with open(unicode_data_file) as unicode_data:
-            for line in csv.reader(unicode_data, delimiter=';'):
-                unicode_id = int(line[0], 16)
+        interval_sps = []
+        interval_lengths = []
+        chars = []
+
+        ranges = list(self._group_ranges(units))
+
+        for element in ranges:
+            interval_length = element[1] - element[0]
+            if interval_length == 0:
+                chars.append(element[0])
+            elif interval_length > self._length_limit:
+                for i in range(element[0], element[1], self._length_limit + 1):
+                    length = self._length_limit if (element[1] - i > self._length_limit) else (element[1] - i)
+                    interval_sps.append(i)
+                    interval_lengths.append(length)
+            else:
+                interval_sps.append(element[0])
+                interval_lengths.append(element[1] - element[0])
 
-                # Skip supplementary planes and ascii chars
-                if unicode_id >= 0x10000 or unicode_id < 128:
-                    continue
+        return interval_sps, interval_lengths, chars
 
-                category = line[2]
+    def read_units(self, file_path, category):
```

Comment:
Maybe we could reorganize this a bit to avoid multiple passes on the same input file. What do you thing about this?: 

```python
def read_units(self, file_path, categories):
    """
    Read the Unicode Derived Core Properties file and extract the ranges
    for the given categories.

    :param file_path: Path to the Unicode "DerivedCoreProperties.txt" file.
    :param categories: A list of category strings to extract from the Unicode file.
    :return: A dictionary each string from the :param categories: is a key and for each
             key list of code points are stored.
    """
    # Create a dictionary in the format: { category[0]: [ ], ..., category[N]: [ ] } 
    units = dict(zip(categories, [[]] * len(categories)))

    with open(file_path, "r") as fp:
        for line in fp:
            # Formats to match:
            #  <HEX>     ; <category> #
            #  <HEX>..<HEX>     ; <category> #
            match = re.match(r"(?P<start>[\dA-F]+)(?:\.\.(?P<end>[\dA-F]+))?\s+; (?P<category>[\w]+) #", line)

            if match and match.group("category") in categories:
                start = int(match.group("start"), 16)
                # if no "end" found use the "start"
                end = int(match.group("end") or match.group("start"), 16)

                matching_code_points = [
                    code_point for code_point in range(start, end + 1) if self.in_range(code_point)
                ]
                units[match.group("category")].extend(matching_code_points)

    return units
```

With this the required categories can be read in a single pass over the input file